packages/expect/src/__tests__/matchers.test.js

```diff
@@ -1615,6 +1615,13 @@ describe('.toMatch()', () => {
   it('escapes strings properly', () => {
     jestExpect('this?: throws').toMatch('this?: throws');
   });
+
+  it('does not maintain RegExp state between calls', () => {
+    const regex = /[f]\d+/gi;
+    jestExpect('f123').toMatch(regex);
+    jestExpect('F456').toMatch(regex);
```

Comment:
I don't really know how I feel about this. As a JS user, I would expect this regex to be stateful and would be surprised that it's re-set by a specific matcher. cc @pedrottimark @SimenB

Reply:
I agree @thymikee, FWIW that's the entire point of having a `global` flag in the first place is to be able to test against multiple matches. The issue reported via #9283 could be solved by removing the `global` from the RegExp since it's doing what it's supposed to do.

That said, #9283 was marked as a bug, because this behavior (specifically with Jest) could also be considered a regression since this was not the intended behavior before v24 was released.

I'm not sure if [this behavior change](https://github.com/facebook/jest/commit/c3a01676b06e9b2372ef7d883256c6745ba3c25c#diff-99739ad11655a4534ce3ab4222dbddbcL687) was intentional or documented somewhere (since it was also a breaking change), or if it was just a regression that needs to be fixed (hence this PR).

Reply:
cc @pedrottimark

Reply:
For what it's worth, I wouldn't expect Jest to change the state of an object, including a regular expression. It seems like it creates confusion and potential for false-positives in cases where the regular expression is being tested against several strings in the same test. If I want to test the `lastIndex` of a regex, I think I would test that directly:

```js
regex.test('f123');
expect(regex.lastIndex).toBe(4);
regex.test('f456');
expect(regex.lastIndex).toBe(0);
```

Reply:
You have my apology that I missed commenting on the issue or pull request before this.

It was my mistake in #8008 not to keep the code path to create a new instance for expected **RegExp** when I changed an expected **string** to match using `includes` method.

Therefore, this PR fixes what I broke, no more and no less.

@wsmd Thanks for your work. If you can merge changes from master and resolve the conflict in `CHANGELOG.md` then this is ready to merge.

There is a related (but separate) documentation chore in `ExpectAPI.md` under `toMatch` to adapt the example in #9283 and in the preceding https://github.com/facebook/jest/pull/9289#discussion_r366056897 so people understand how to write tests that are independent of RegExp state or intentionally testing RegExp state.

Reply:
Thanks @pedrottimark! The merge conflict is now resolved. 👍